### PR TITLE
[persist] Perform expiry directly instead of passing it as maintenance

### PR DESCRIPTION
**Pull request by bkirwi**
### Motivation

When discussing #17150, we discovered that our current approach to expiry can result in **O(n^2)** command applications, where **n** is the number of handles expiring for a single shard at once. (The call to `apply_unbatched_cmd` will re-discover and relaunch expiry for the remaining **n-1** expiring handles, etc.)

Expiry is cheap -- we already scan through the list of handles once per command, so we can just drop the expired ones immediately instead of routing them through our maintenance machinery. This means expiry no longer places additional load on the state machine. (It also avoids any races where a lease is renewed after we've already decided to expire it; I think those are harmless but it's nice to have less to think about.)

### Tips for reviewer

Originally, I was hoping to expire handles based on the walltime of the state. However! This could lead to bad behaviour when clocks skew... if one clock is ahead for, say, an hour, it can advance the walltime to an hour ahead of the current time, and then any leases that are registered within the next 45 minutes will be expired immediately. I have ideas about how to avoid this - choosing the lease start time based on the walltime, etc. - but that seems like a larger change with much smaller marginal value, so I went with this simpler thing that keeps semantics closer to the existing code.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by bkirwi** [comment]
I haven't been able to reproduce the "run maintenance everywhere" test failures when these changes are also merged in, similar to #17150. That suggests (but does not prove!) that the rapid blowup in memory was many copies of state triggered by a flurry of expiring handles. (And it's easy to see how doing maintenance more often or aggressively could have made that problem worse.)

File: src/persist-client/src/internal/state.rs [hunk 1]
```diff
@@ -1154,31 +1177,6 @@ where
...
-    ) -> (Vec<LeasedReaderId>, Vec<WriterId>) {
-        let mut readers = Vec::new();
-        for (reader, state) in self.collections.leased_readers.iter() {
-            let time_since_last_heartbeat_ms =
-                now_ms.saturating_sub(state.last_heartbeat_timestamp_ms);
-            if time_since_last_heartbeat_ms > state.lease_duration_ms {
-                readers.push(reader.clone());
-            }
-        }
-        // critical_readers don't need forced expiration (in fact, that's the
```
Comment (danhhz) [inline src/persist-client/src/internal/state.rs:1169]: just making sure we dropped this comment about critical readers on purpose (fine, if so)

**Review by danhhz — APPROVED** [review]
such a great idea, I love it!

the one reservation I have is the idea of a silent no-op registration where some reader or writer expires itself. can we please detect that case and put a `tracing::error!` on it so it at least shows up in sentry if it ever happens?

**Comment by danhhz** [comment]
> I haven't been able to reproduce the "run maintenance everywhere" test failures when these changes are also merged in, similar to https://github.com/MaterializeInc/materialize/pull/17150. That suggests (but does not prove!) that the rapid blowup in memory was many copies of state triggered by a flurry of expiring handles. (And it's easy to see how doing maintenance more often or aggressively could have made that problem worse.)

oh that's _very_ exciting

**Comment by bkirwi** [comment]
Added a couple more comments and some `error!` logs that complain if a handle isn't present in state immediately after registering it. (It's not necessarily evidence of a bug if it happens, and the equivalent thing is possible in today's maintenance... but it does seem like something that ought to be extremely rare and thus worth tracking.) This branch also passed CI with those new logs converted to panics, which is kind of nice.

[Here's our nightly run.](https://buildkite.com/materialize/nightlies/builds/1709) The failing tests seem to seem to be known-failing, and the flaky tests seem to be known-flaky, but it's still not a very comforting picture. I'll see if I can get a cleaner run when some of the known issues are resolved, assuming that gets sorted out in the next day or two.

**Comment by danhhz** [comment]
FYI I've started rebasing all my PRs on top of the latest commit that passed nightly for exactly that reason. works okay as long as there aren't any more recent merge conflicts

File: src/persist-client/src/internal/state.rs (hunk 1 above)
Comment (bkirwi) [reply src/persist-client/src/internal/state.rs:1169]: I didn't have any particular intent one way or the other, so added it back for now; if we decide to clear out comments that can be a separate change.

**Comment by bkirwi** [comment]
Sensible! Looks like we have a green nightly when rebased on last week's `main`: https://buildkite.com/materialize/nightlies/builds/1717

**Pull request merged**